src/coreclr/nativeaot/Runtime/wasm/FinalizerHelpers.SingleThreaded.cpp

```diff
@@ -0,0 +1,76 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+//
+// Unmanaged portion of finalization implementation for a single-threaded environment.
+// Currently, only supports explicit finalization via WaitForPendingFinalizers.
+//
+#include "common.h"
+#include "gcenv.h"
+#include "gcheaputilities.h"
+#include "gcrhinterface.h"
+
+#include "thread.h"
+#include "threadstore.h"
+#include "threadstore.inl"
+#include "thread.inl"
+
+// Finalizer method implemented by the managed runtime.
+extern "C" __cdecl void RhpProcessFinalizersAndReturn();
+
+static void ProcessFinalizersAndReturn()
+{
+    static bool s_finalizing = false;
+
+    // Recursive wait on finalization is a no-op.
+    if (!s_finalizing)
+    {
+        s_finalizing = true;
+        RhpProcessFinalizersAndReturn();
+        s_finalizing = false;
+    }
+}
+
+bool RhInitializeFinalization()
+{
+    return true;
+}
+
+// This method is called at the end of GC in case finalizable objects were present.
+void RhEnableFinalization()
+{
+    // TODO: Implement automatic finalization. Cannot just call "ProcessFinalizersAndReturn"
+    // here as it will deadlock the GC.
+}
+
+EXTERN_C NATIVEAOT_API void __cdecl RhWaitForPendingFinalizers(UInt32_BOOL allowReentrantWait)
+{
+    // Must be called in cooperative mode as "ProcessFinalizersAndReturn" RPIs back into managed.
```

Comment:
Should this comment be other way around? The assert is verifying that it is not called in cooperative mode.

Reply:
> Should this comment be other way around?

Yep. Fixed it.